src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs

```diff
@@ -5,130 +5,188 @@
+            ProtocolDiagnostics.HttpEvent += ProtocolDiagnostics_HttpEvent;
+            ProtocolDiagnostics.ResourceEvent += ProtocolDiagnostics_ResourceEvent;
+            ProtocolDiagnostics.NupkgCopiedEvent += ProtocolDiagnostics_NupkgCopiedEvent;
             _parentId = parentId;
+            _actionName = actionName;
 
             // Multiple sources can use the same feed url. We can't know which one protocol events come from, so choose any.
-            _sources = new Dictionary<string, PackageSource>();
+            _sources = new Dictionary<string, SourceRepository>();
             foreach (var source in sources)
             {
-                _sources[source.Source] = source;
+                _sources[source.PackageSource.Source] = source;
+            }
+
+            _knownSources = new HashSet<string>();
+            foreach (var source in _sources.Keys)
+            {
+                _knownSources.Add(source);
             }
         }
 
-        private void ProtocolDiagnostics_Event(ProtocolDiagnosticEvent pdEvent)
+        private void ProtocolDiagnostics_ResourceEvent(ProtocolDiagnosticResourceEvent pdEvent)
         {
-            AddAggregateData(pdEvent, _data);
+            AddResourceData(pdEvent, _data, _resourceStringTable, _knownSources);
         }
 
-        internal static void AddAggregateData(ProtocolDiagnosticEvent pdEvent, ConcurrentDictionary<string, Data> allData)
+        internal static void AddResourceData(
+            ProtocolDiagnosticResourceEvent pdEvent,
+            ConcurrentDictionary<string, Data> allData,
+            ConcurrentDictionary<string, ConcurrentDictionary<string, string>> resourceStringTable,
+            HashSet<string> sources)
         {
+            if (!sources.Contains(pdEvent.Source))
+            {
+                return;
+            }
+
+            var resourceMethodNameTable = resourceStringTable.GetOrAdd(pdEvent.ResourceType, t => new ConcurrentDictionary<string, string>());
+            var resourceTypeAndMethod = resourceMethodNameTable.GetOrAdd(pdEvent.Method, m => pdEvent.ResourceType + "." + m);
+
             var data = allData.GetOrAdd(pdEvent.Source, _ => new Data());
+            lock (data._lock)
+            {
+                if (data.Resources.TryGetValue(resourceTypeAndMethod, out var t))
+                {
+                    data.Resources[resourceTypeAndMethod] = (t.count + 1, t.duration + pdEvent.Duration);
+                }
+                else
+                {
+                    data.Resources[resourceTypeAndMethod] = (1, pdEvent.Duration);
+                }
+            }
+        }
 
-            var resourceData = pdEvent.Url.OriginalString.EndsWith(".nupkg", StringComparison.OrdinalIgnoreCase)
-                ? data.Nupkg
-                : data.Metadata;
+        private void ProtocolDiagnostics_HttpEvent(ProtocolDiagnosticHttpEvent pdEvent)
+        {
+            AddHttpData(pdEvent, _data, _knownSources);
```

Comment:
Not following the purpose of having _data and _knownSources as parameters.
If they are properties, and the methods are non-static, why not just reference them directly?

These are also private/internal methods, so it's not like someone has the ability to pass in their own collections here.

Reply:
The class constructor makes global state changes (subscribes to a static event). I don't want tests to change or use global state, particularly so if tests run in parallel, there are no flaky failures. So, all the instance methods call static methods with the business logic, so that tests can call the static methods directly, without affecting any static values.